### ISSUE TITLE
Change connection logic for Sunbros to preferr lower SL

### ISSUE THREAD

**Issue by Chronial** [Wulf2k/DaS-PC-MPChan#111]
I tried changing the connection logic for my sunbro so it connects to people at the bottom of my summoning range instead of close to my SL. My subjective impression is that that increased my summoning rate. I would also say that that makes sense in general – lower SL players are more likely to need help with a boss, right?

@Wulf2k what do you think? I would only do this if the player is a Warrior of Sunlight, because I think we can take this as a clear sign that they want to be summoned to help other players. I'm not sure whether we should treat Princess Guard the same.

**Comment by Wulf2k** [comment]
I'd agree with this.  Sunbros should prefer lower SL.

I have no problem extending it to PG, but I wouldn't care if it wasn't.  We should probably also aim to make the wiki accurate, and prioritize the other co-op covenants when possible.

The co-op covenants would only add each other thereby reducing their invasions, but not completely eliminating due to the darkwraiths adding the co-op players themselves.

**Comment by 0int** [comment]
> The co-op covenants would only add each other thereby reducing their invasions, but not completely eliminating due to the darkwraiths adding the co-op players themselves.

This will drastically limit the usage of PG's healing miracles in random PvP as well as the possibility of assisting players who need some coop help in PvE but don't want to abandon a PvP/Chaos covenant for the sake of preserving their covenant level. Not sure if it is a good thing. Probably doing this to the Way of White would be enough (and finally it will make sense to join this covenant).

Think through this sevenfold plz.

**Comment by 0int** [comment]
Then again, implementing this logic for the sunbro covenant will turn out so sunbros will get invaded almost only by the lower level invaders, not equals or slightly stronger (and of course almost no darkmoon police). Am I right?

**Comment by Chronial** [comment]
> We should probably also aim to make the wiki accurate, and prioritize the other co-op covenants when possible.

Imho basing connection decisions on DSCM-Only data is not such a great idea, as it reduces the pool of players to about 1/4. But I think "punishing" nodes based on DSCM data is a viable strategy (this is already done for forest invaders for example – they will not connect to other forest invaders). So we could make Coop covenanters less likely to connect to PVP covenanters. I like that idea.

> This will drastically limit the usage of PG's healing miracles in random PvP

What is the use of those miracles in random PVP? Aren't they only usefull in coop?

> as well as the possibility of assisting players who need some coop help in PvE but don't want to abandon a PvP/Chaos covenant 

I think there should be a price for beeing in those covenants. Each covenant offers benefits, but you also have to pay a price.

> Then again, implementing this logic for the sunbro covenant will turn out so sunbros will get invaded almost only by the lower level invaders, not equals or slightly stronger (and of course almost no darkmoon police). Am I right?

Not really. You have to remember that this only affects one "kind" of connection. There are basically three kinds of connections:

1. Created by the `ingame/steam matchmaking`: As far as we know these are basically random.
2. Created by `your DSCM` to somebody else: These follow the logic we are discussing here
3. Created by `somebody else's DSCM` to you: These also follow the logic here, but based on **their** interests (including their covenant).

Darkmoon should for example not be effected by this at all. As a Darkmoon Wraith, the probabilty that a sinner just randomly (`1` or `3`) connects to you is quite low. You will get your sinners by actively picking them out (`2`).

**Comment by 0int** [comment]
> Not really. You have to remember that this only affects one "kind" of connection. There are basically three kinds of connections
I see. Thanks for explainations. Now I don't think the other issues I pointed out will be a problem but still let me explain:

> What is the use of those miracles in random PVP? Aren't they only usefull in coop?
You see, coop isnt only a PvE thing. As a fellow darkwraith I find it amusing to invade cooping people, or do 2v2/1v2 fights in darkroots as a catbro. 

>  think there should be a price for beeing in those covenants. Each covenant offers benefits, but you also have to pay a price.
I can't disagree with you here, but I find it ironic that people who don't care for PvP and stick to Chaos Covenant only to keep Solaire alive will have to suffer a lack of Sunbros help throughout the game.

I dare to suggest the next thing:
1) You set the Ways of White connectivity logic to prefer connections with other coops covenants, and set darkwraith/gravelord covenants (darkmoons and catbros are under a question here due to obvious lore reasons) logic to ignore WoW members (as well as WoW will ignore darkwraiths themselves) so people can stay mostly out of PvP if they don't want it. That way they wouldn't occupy a node slot by staying hollow and being a dummy node which could never be invaded anyway `(btw, consider marking people who stay hollowed for a long time as poor nodes?)`. It would fix this useless covenant according to a wiki description and make it offering some unique benefits at last.
2) You dont mess much with other coop covenants' connectivity logic, so they will still stay in a good touch with common invasions to withstand them in jolly cooperation, and so Sunbros can offer help to everyone, not to their sect only.
Btw here is a thing: sunbrethren offer random help, but they rarely really have a need of random help themselves (they mostly hang around their signs), so making them to prioritize their connectivity on their own covenant doesn't seem to have an effective impact on matchmaking for me

Once again, about your suggestion of attuning the connection logic for Sunbro to prioritize the lower levels. Sure, it is ok for sunbro to get summoned more this way, but it is not much fun for host to have an overpowered phantom who does all job for him ez-pz. It feels more like patronizing than a jolly cooperation. Sure, that depends on builds and people mindset, but, do you really need to actively support that?

**Comment by Chronial** [comment]
> set darkwraith/gravelord covenants (darkmoons and catbros are under a question here due to obvious lore reasons) logic to ignore WoW members

If you are sinner, you need to repent, independent of your covenant ;). This seems a bad idea – most Beginners will be in Way of White, and they should not be exempt from this part of the game.

> being a dummy node which could never be invaded anyway 

Just because somebody isn't looking for pvp doesn't mean you can't invade them ;)

> but it is not much fun for host to have an overpowered phantom who does all job for him ez-pz

This is actually *way* more dependent on gear and skill than on SL. SL 100 vs 120 is not that big a difference – it's your gear that will make the big difference. So I think that is an acceptable trade-off for getting Sunbros more summons. And once again remember that the other player will connect to nodes of the same SL, which might also be a Sunbro. So this change will mostly change what happens to sunbros, way less what happens to everybody else.

The DSCM-using sunbros are 5% of the DSCM-Net, so they won't distort general coop to a general "the person I'm summoning is higher SL than me" thing anyway.

> Btw here is a thing: sunbrethren offer random help, but they rarely really have a need of random help themselves

This one of the reason for this change: The other sunbros chilling near their sign are probably closer to your SL, whereas new players entering the area are probably below it.

**Comment by 0int** [comment]
> Just because somebody isn't looking for pvp doesn't mean you can't invade them ;)

You are right, but the thing is - why would you want a 24/7 hollowed guy to stay in your node pool for long?
I mean - recently I was about to invade in Tomb of Giants, but DSCM node rooster shown me that every player around was hollowed, so I had a little hope of getting interactions.

> SL 100 vs 120 is not that big a difference

True, but SL 9 VS SL 20 means quite a difference, due the ranges of available gear.

> This one of the reason for this change: The other sunbros chilling near their sign are probably closer to your SL, whereas new players entering the area are probably below it.

Thats quite a daring assumption. Good players just level-up as they progress naturaly, mostly even trying not to get overlevelled. New/bad players (those who would most likely need a hand) are mostly of higher level due to grinding for souls, so the SL spread is pretty wide.
________

Look, you don't have to persuade me in anything. It's you guys who do the work here, not me. I just was a bit worried on how your mod would change the matchmaking in general, so I expressed my concerns here. If you are sure everything will be ok then I trust you and I am sure everything will end up in a good way. 
Thank you guys for all the good work you do.

**Comment by Wulf2k** [comment]
0int, I suppose I can't speak for chronial but I personally find opposing viewpoints interesting.  It's easy to make an assumption and then carry on as if it were fact if nobody challenges it.

On the matchmaking preferring humans:  Yes, it would be best from the individual's point of view if he only connected to humans, but consider this scenario.  There are 50 people in Undead Burg.  5 of them are human.  Each of those 45 hollows is continuously attempting to connect to those 5 humans, and those 5 humans have a maximum of 20 nodes.  What happens?  I dunno, but I'm guessing it's not ideal.

Also, the DSCM-Net list is kind of a snapshot in time.  I don't know the rate of turnover, but due to the nature of the game, those 5 humans will most likely stop being human at some point in the relatively near future, and a bunch of those previous hollows will become human themselves.  Should DSCM just be like an auto-turret, firing an overwhelming mass of connection attempts at whoever is currently human, or should it spread those connections out and just hope for some phantom churn?

On the covenant matching:  I dunno yet, I still need caffeine.

**Comment by hydra-chan** [comment]
Have boss flags been implemented already? Sun covenanters should prefer people with a boss flag set. As 0int mentioned, newer players may decide they want to amass souls for stats, so players who need help aren't always going to be low SL characters. It's not a very good indicator of who needs help, especially given that twinks also exist as low SL chars. I think it's best to randomly select people who for sure probably need help (have a positive boss flag) and have a quick turnover on players who have a clearer boss flag.

**Comment by 0int** [comment]
Here is a thing: those players shouldn't be purged as soon as the flag is clear. It would be nice to let Sunbro assist this same player in the next location or treacherously invade him there.

**Comment by Chronial** [comment]
> 0int, I suppose I can't speak for chronial but I personally find opposing viewpoints interesting. It's easy to make an assumption and then carry on as if it were fact if nobody challenges it.

This :). Designing distributed systems is hard, viewpoints help.

Regarding connecting to humans: For coop this is completely useless. We only know for sure that a player is human *and* looking for coop for the duration it takes to walk from the bonfire to the summon signs, which is maybe 10 seconds. For the reasons Wulf2k stated, we can't do anything with that.

For PVP that might be helpful, but as Wulf already said – that would cause everybody to swoop down on you as soon as your humanity is reported to DSCM-Net.

> Have boss flags been implemented already? Sun covenanters should prefer people with a boss flag set.

No, they have not been implemented yet, but they would be very useful for matchmaking, now that we have a sizeable amount of players in DSCM-Net. Feel free to implement it :wink:.

> … have a quick turnover on players who have a clearer boss flag.

and

> those players shouldn't be purged as soon as the flag is clear

At the moment we are very careful with disconnecting nodes – it only happens after a node was clearly useless (no mp interaction possible because of SL and area) for the last five minutes. Being more aggressive might interfere with normal matchmaking.

**Comment by Wulf2k** [comment]
Disconnecting people with cleared boss flags would definitely impact PvP.

**Comment by Wulf2k** [comment]
Also, boss flags are on the list of things I'm actively poking at whenever I have time.

I suck at prioritizing though, so who knows when I'll get that finished up....  Namecrashey stuff should be done before the boss flags.  I'm bouncing around depending on whichever strikes my fancy at the time.
